src/Functions/FunctionsStringHashFixedString.cpp

```diff
@@ -34,173 +35,159 @@ extern "C" {
 
 namespace DB
 {
+
 namespace ErrorCodes
 {
-extern const int ILLEGAL_COLUMN;
-extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int ILLEGAL_COLUMN;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int OPENSSL_ERROR;
+    extern const int LOGICAL_ERROR;
 }
 
 
 #if USE_SSL
+using EVP_MD_CTX_ptr = std::unique_ptr<EVP_MD_CTX, decltype(&EVP_MD_CTX_free)>;
+
+template <typename ProviderImpl>
+class OpenSSLProvider
+{
+public:
+    static constexpr auto name = ProviderImpl::name;
+    static constexpr auto length = ProviderImpl::length;
+
+    OpenSSLProvider() : ctx_template(EVP_MD_CTX_new(), &EVP_MD_CTX_free)
+    {
+        if (!ctx_template)
+            throw Exception(ErrorCodes::OPENSSL_ERROR, "EVP_MD_CTX_new failed: {}", getOpenSSLErrors());
+
+        if (EVP_DigestInit_ex(ctx_template.get(), ProviderImpl::provider(), nullptr) != 1)
+            throw Exception(ErrorCodes::OPENSSL_ERROR, "EVP_DigestInit_ex failed: {}", getOpenSSLErrors());
+    }
+
+    void apply(const char * begin, size_t size, unsigned char * out_char_data)
+    {
+        if (!ctx_template)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "No context provided");
+
+        thread_local EVP_MD_CTX_ptr ctx(EVP_MD_CTX_new(), &EVP_MD_CTX_free);
+        if (EVP_MD_CTX_copy_ex(ctx.get(), ctx_template.get()) != 1)
```

Comment:
For my understanding: We copy something from the template context to the thread context once per `apply()` invocation?